### PR TITLE
Add install/build scripts for each package.json

### PR DESCRIPTION
**Pull request by ebidel**
R: all

Cherry-picking some stuff from https://github.com/GoogleChrome/lighthouse/pull/1169.

This PR unifies some of the build scripts into the main package.json and uses npm's `prepublish` to build all modules when users `npm install` lighthouse and before we publish. Should help us prevent publishing errors like https://github.com/GoogleChrome/lighthouse/issues/1405 and simplifies deployment process.

### CODE REVIEW

**Comment by wardpeet** [comment]
Looking good! 👍 
I'm going to dive into old commits why we went away from this approach in the first place.

File: package.json [hunk 1]
```diff
@@ -11,6 +11,10 @@
     "node": ">=5"
```
Comment (wardpeet) [inline package.json]: Can we set it to 4 because we still support it 😄

File: lighthouse-viewer/package.json [hunk 2]
```diff
@@ -5,6 +5,7 @@
     "node": ">=5"
```
Comment (wardpeet) [inline lighthouse-viewer/package.json]: Can we set it to 4 because we still support it 😄

File: package.json (hunk 1 above)
Comment (ebidel) [inline package.json]: is there a way to specify node 4 but with the `--harmony` flag?

Comment (wardpeet) [inline package.json]: good point, it is not. so >=5 is better! 👍

File: lighthouse-viewer/package.json (hunk 2 above)
Comment (wardpeet) [inline lighthouse-viewer/package.json]: good point, it is not. so >=5 is better! 👍

File: lighthouse-cli/package.json [hunk 3]
```diff
@@ -1,6 +1,7 @@
 {
   "main": "index.js",
   "scripts": {
+    "prepublish": "npm run build",
```
Comment (brendankenny) [inline lighthouse-cli/package.json]: this is basically the purpose of `prepublish`, but like with `install` below, it feels weird to me to build everything, test it for release, then run `npm publish` and have everything rebuilt (albeit theoretically to the same thing just tested) before being pushed. We can try it out and see if there are issues, I guess.

File: lighthouse-cli/compiled-check.js [hunk 4]
```diff
@@ -6,7 +6,7 @@ const path = require('path');
 module.exports = function(filename) {
   if (!fs.existsSync(path.join(__dirname, filename))) {
     console.log('Oops! Looks like the CLI needs to be compiled. Please run:');
-    console.log('   cd lighthouse-cli && npm install && npm run build');
+    console.log('   cd lighthouse-cli && npm install');
```
Comment (brendankenny) [inline lighthouse-cli/compiled-check.js]: wouldn't we not see this anymore, as all installs (either `npm install lighthouse` (which has these files included) and `npm install` inside the cloned repo) will have the compiled files?

**Comment by brendankenny** [comment]
we may want to fix #1327 before landing this

File: package.json [hunk 5]
```diff
@@ -11,6 +11,10 @@
...
   },
   "scripts": {
+    "prepublish": "npm run install-cli && npm run install-extension && npm run install-viewer",
```
Comment (ebidel) [inline package.json]: We've been led astray with `prepublish`:

screen shot 2017-01-09 at 6 15 26 pm

In light of that deprecation, let's do these on `npm install` instead?

For  non-Travis node v4 users, we could create`setNodeHarmony.sh` that adds the `--harmony` flag if needed and run it in Travis and before the extension builds itself.

File: lighthouse-cli/compiled-check.js (hunk 4 above)
Comment (ebidel) [inline lighthouse-cli/compiled-check.js]: In theory yes, but I was afraid to remove it.

File: package.json (hunk 5 above)
Comment (brendankenny) [inline package.json]: ah, very interesting. SG. I hadn't seen that change, only the controversy back in 2013 when it started to run `prepublish` on `install` :) We should hunt down the current discussions on this

**Comment by ebidel** [comment]
1000 commits later...

PTAL. Removed the `prepublish` scripts b/c it's will no longer be called in npm5 when running `npm install` locally. However, I've kept the `prepublish` for the cli because it's important we don't miss that one again when publishing to npm.`tsc` is also super fast.

After removals, this isn't a whole lot better than before, but it's a good step towards a centralized workflow for building/managing the modules and simplifies some of our stuff.

**Comment by wardpeet** [comment]
what about postinstall? I liked the idea that everything is ready to go on a single install.
Prepublish is nice for cli indeed so we never forget to pack it and send it to npm. :1st_place_medal:

**Comment by ebidel** [comment]
`install`/`postinstall` are discouraged: https://docs.npmjs.com/misc/scripts#best-practices

`install` is also not ideal for normal users that install lighthouse using `npm install -g lighthouse`.

**Comment by wardpeet** [comment]
mmh correct :)

File: package.json [hunk 6]
```diff
@@ -11,6 +11,14 @@
     "node": ">=5"
   },
   "scripts": {
+    "install-all": "npm run install-cli && npm run install-extension && npm run install-viewer",
+    "install-cli": "npm --prefix ./lighthouse-cli install ./lighthouse-cli",
+    "install-extension": "npm --prefix ./lighthouse-extension install ./lighthouse-extension",
+    "install-viewer": "npm --prefix ./lighthouse-viewer install ./lighthouse-viewer",
```
Comment (paulirish) [inline package.json]: this and the other `installs` dont need to specify the path, afaict.

can be just `npm install -C ../lighthouse-viewer`

Comment (paulirish) [inline package.json]: (I prefer the `-C` to `--prefix` as its more git-like)

File: package.json [hunk 7]
```diff
@@ -11,6 +11,14 @@
...
+    "build-all": "npm run build-cli && npm run build-extension && npm run build-viewer",
+    "build-cli": "cd lighthouse-cli && npm run build",
```
Comment (paulirish) [inline package.json]: lets change these to also be `npm run build -C ./lighthouse-cli` etc

File: package.json [hunk 8]
```diff
@@ -11,6 +11,14 @@
     "node": ">=5"
   },
   "scripts": {
+    "install-all": "npm run install-cli && npm run install-extension && npm run install-viewer",
```
Comment (paulirish) [inline package.json]: most of us use yarn.. are these scripts only for bots?

Comment (ebidel) [inline package.json]: yes + convenience. They'll still work with yarn.

File: package.json (hunk 6 above)
Comment (ebidel) [inline package.json]: done

File: package.json (hunk 7 above)
Comment (ebidel) [inline package.json]: done

**Comment by ebidel** [comment]
PTAL

**Comment by ebidel** [comment]
@paulirish merge?

File: lighthouse-cli/compiled-check.js [hunk 9]
```diff
@@ -6,7 +6,7 @@ const path = require('path');
 module.exports = function(filename) {
   if (!fs.existsSync(path.join(__dirname, filename))) {
     console.log('Oops! Looks like the CLI needs to be compiled. Please run:');
-    console.log('   cd lighthouse-cli && npm install && npm run build');
+    console.log('   npm run install-cli');
```
Comment (brendankenny) [inline lighthouse-cli/compiled-check.js]: doesn't this have to be built again?

File: lighthouse-cli/package.json (hunk 3 above)
Comment (brendankenny) [inline lighthouse-cli/package.json]: do we want prepublish anymore?

File: package.json (hunk 8 above)
Comment (brendankenny) [inline package.json]: not all of us use `yarn` :P

File: lighthouse-cli/compiled-check.js (hunk 9 above)
Comment (ebidel) [inline lighthouse-cli/compiled-check.js]: The cli has: `"prepublish": "npm run build",`. It'll be built on `npm install`

Comment (brendankenny) [inline lighthouse-cli/compiled-check.js]: I thought it was no longer going to do that in npm 5?

File: lighthouse-cli/package.json (hunk 3 above)
Comment (ebidel) [inline lighthouse-cli/package.json]: See https://github.com/GoogleChrome/lighthouse/pull/1439#issuecomment-271480016

File: lighthouse-cli/compiled-check.js (hunk 9 above)
Comment (ebidel) [inline lighthouse-cli/compiled-check.js]: Right. I thought it would still be a good idea to have this around until that happens...given our history :\

**Pull request merged**